lib/factory_bot/definition_proxy.rb

```diff
@@ -230,8 +274,11 @@ def trait(name, &block)
     #     those traits. When this argument is not provided, factory_bot will
     #     attempt to get the values by calling the pluralized `attribute_name`
     #     class method.
-    def traits_for_enum(attribute_name, values = nil)
-      @definition.register_enum(Enum.new(attribute_name, values))
+    def traits_for_enum(attribute_name, values = nil, _prefix: false, _suffix: false, **values_as_hash)
```

Comment:
Supporting both hash and keyword arguments is making this a bit complicated. I don't think we need to match the Rails `enum` method exactly here, since the use cases are a bit different (for example, I think passing the values to `traits_for_enum` as keyword arguments would be fairly uncommon, and it's also not something we've tested).

I'd rather accept a single values argument (like a hash or array), and keep the prefix and suffix keyword arguments separate. That also allows us to get rid of the ugly underscores.

So that would look like:

```rb
    def traits_for_enum(attribute_name, values = nil, prefix: false, suffix: false)
      @definition.register_enum(Enum.new(attribute_name, values, prefix: prefix, suffix: suffix))
    end
```

All of our tests still pass with this change (assuming we update the _prefix and _suffix keyword arguments to remove the underscores).

Reply:
from what i can see, calling
```
statuses = {queued: 0, started: 1, finished: 2}
traits_for_enum(:status, statuses)
```
will fail with this definition
```
traits_for_enum(attribute_name, values = nil, prefix: false, suffix: false)
```
the error I'm getting is
```
ArgumentError: unknown keywords: :queued, :started, :finished
```
My understanding here is ruby is converting the hash to keyword on the way in due to the prefix/suffix keywords.

I can continue to investigate this further to get the current expectation to work (i'm assuming my understanding of ruby's keywords is the issue here, not ruby itself)

The other option I can see is having another explicit "options" hash as a third argument.
```
def traits_for_enum(attribute_name, values = nil, options = {})
  @definition.register_enum(Enum.new(attribute_name, values, prefix: options.fetch(:prefix, false), suffix: options.fetch(:suffix, false)))
end
```
which will be called as:
```
traits_for_enum :status, {queued: 0, started: 1, finished: 2}, {prefix: true}
```

any preferences which way to go?

Reply:
the issue here is not necessarily that key names are needed, but more that a third argument is needed for setting options, but the traits_for_enum allows for values to be nil (in which case the second arg passed becomes options, not values).
if traits_for_enum is invoked with values being passed in, a third arg with options works fine, but if values is not passed in, the "options" now are the second arg, meaning values (was meant to default to nil) and options (was meant to be passed in) are both jeopardised.

i can think of 2 options here,
1: the code update I've made, which looks good from an api usage point of view, but means traits_for_enum becomes messy, as it has to try to determine on the fly what the intention of the second arg is.

2: if "options" are being used, values is a key name too (eg: `traits_for_enum(:foo, values: [:a, :b], prefix: true, suffix: true)`, which means the consumer of the api has to change their code (only when setting options, which are a new concept so all current usage continues to work) but will need to migrate over when wanting to set prefix/suffix.